ICE: No `move_errors` should be allowed in MIR borrowck

I get an ICE when I try to build [evolvim](https://github.com/syberant/evolvim):
```
error: internal compiler error: No `move_errors` should be allowed in MIR borrowck
   --> evolvim-lib\./src/lib\neat\genome\mod.rs:100:13
    |
100 |             InputType::Eye(EYE[2]).clone(),
    |             ^^^^^^^^^^^^^^^^^^^^^^

error: internal compiler error: No `move_errors` should be allowed in MIR borrowck
  --> evolvim-lib\./src/lib\neat\genome\mod.rs:99:13
   |
99 |             InputType::Eye(EYE[1]).clone(),
   |             ^^^^^^^^^^^^^^^^^^^^^^

error: internal compiler error: No `move_errors` should be allowed in MIR borrowck
  --> evolvim-lib\./src/lib\neat\genome\mod.rs:98:13
   |
98 |             InputType::Eye(EYE[0]).clone(),
   |             ^^^^^^^^^^^^^^^^^^^^^^

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src\librustc_errors\lib.rs:357:17
stack backtrace:
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: <std::io::IoSliceMut as core::fmt::Debug>::fmt
   3: std::panicking::take_hook
   4: std::panicking::take_hook
   5: rustc_driver::report_ice
   6: std::panicking::rust_panic_with_hook
   7: <rustc_errors::snippet::Style as core::fmt::Debug>::fmt
   8: <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop
   9: rustc_driver::pretty::print_after_hir_lowering
  10: rustc_driver::pretty::print_after_hir_lowering
  11: rustc_driver::pretty::print_after_hir_lowering
  12: rustc_driver::pretty::print_after_hir_lowering
  13: <env_logger::filter::inner::Filter as core::fmt::Display>::fmt
  14: <env_logger::filter::inner::Filter as core::fmt::Display>::fmt
  15: <env_logger::filter::inner::Filter as core::fmt::Display>::fmt
  16: _rust_maybe_catch_panic
  17: rustc_driver::pretty::print_after_hir_lowering
  18: ZN244_$LT$std..error..$LT$impl$u20$core..convert..From$LT$alloc..string..String$GT$$u20$for$u20$alloc..boxed..Box$LT$dyn$u20$std..error..Error$u2b$core..marker..Sync$u2b$core..marker..Send$GT$$GT$..from..StringError$u20$as$u20$core..fmt..Display$GT$3fmt17
  19: std::sys::windows::thread::Thread::new
  20: BaseThreadInitThunk
  21: RtlUserThreadStart
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.
```
I tried stable, beta and nightly. All ended with an ICE.
```
>rustc --version --verbose
rustc 1.42.0 (b8cedc004 2020-03-09)
binary: rustc
commit-hash: b8cedc00407a4c56a3bda1ed605c6fc166655447
commit-date: 2020-03-09
host: x86_64-pc-windows-msvc
release: 1.42.0
LLVM version: 9.0
```

Looks like this compiles but it shouldn't:
```rust
struct S;

fn foo() {
    &([S][0],);
}
```
Dumping the MIR for this you get:
```rust
fn  foo() -> () {
    let mut _0: ();                      // return place in scope 0 at src/lib.rs:3:10: 3:10
    let _1: &(S,);                       // in scope 0 at src/lib.rs:4:5: 4:15
    let mut _2: &(S,);                   // in scope 0 at src/lib.rs:4:5: 4:15

    bb0: {
        StorageLive(_1);                 // bb0[0]: scope 0 at src/lib.rs:4:5: 4:15
        _2 = const foo::promoted[0];     // bb0[1]: scope 0 at src/lib.rs:4:5: 4:15
                                         // ty::Const
                                         // + ty: &(S,)
                                         // + val: Unevaluated(DefId(0:5 ~ playground[4e22]::foo[0]), [], Some(promoted[0]))
                                         // mir::Constant
                                         // + span: src/lib.rs:4:5: 4:15
                                         // + literal: Const { ty: &(S,), val: Unevaluated(DefId(0:5 ~ playground[4e22]::foo[0]), [], Some(promoted[0])) }
        _1 = _2;                         // bb0[2]: scope 0 at src/lib.rs:4:5: 4:15
        StorageDead(_1);                 // bb0[3]: scope 0 at src/lib.rs:4:15: 4:16
        return;                          // bb0[4]: scope 0 at src/lib.rs:5:2: 5:2
    }
}

promoted[0] in  foo: &(S,) = {
    let mut _0: &(S,);                   // return place in scope 0 at src/lib.rs:4:5: 4:15
    let mut _1: (S,);                    // in scope 0 at src/lib.rs:4:6: 4:15
    let mut _2: S;                       // in scope 0 at src/lib.rs:4:7: 4:13
    let mut _3: [S; 1];                  // in scope 0 at src/lib.rs:4:7: 4:10
    let mut _4: S;                       // in scope 0 at src/lib.rs:4:8: 4:9
    let mut _5: usize;                   // in scope 0 at src/lib.rs:4:11: 4:12

    bb0: {
        _3 = [move _4];                  // bb0[0]: scope 0 at src/lib.rs:4:7: 4:10
        _5 = const 0usize;               // bb0[1]: scope 0 at src/lib.rs:4:11: 4:12
                                         // ty::Const
                                         // + ty: usize
                                         // + val: Value(Scalar(0x0000000000000000))
                                         // mir::Constant
                                         // + span: src/lib.rs:4:11: 4:12
                                         // + literal: Const { ty: usize, val: Value(Scalar(0x0000000000000000)) }
        _2 = move _3[_5];                // bb0[2]: scope 0 at src/lib.rs:4:7: 4:13
        (_1.0: S) = move _2;             // bb0[3]: scope 0 at src/lib.rs:4:6: 4:15
        _0 = &_1;                        // bb0[4]: scope 0 at src/lib.rs:4:5: 4:15
        return;                          // bb0[5]: scope 0 at src/lib.rs:4:5: 4:15
    }
}
```

Presumably, what should happen here is:
```
error[E0508]: cannot move out of type `[S; 1]`, a non-copy array
 --> src/lib.rs:4:6
  |
4 |    &([S][0],);
  |      ^^^^^^
  |      |
  |      cannot move out of here
  |      move occurs because value has type `S`, which does not implement the `Copy` trait
```

The `delay_span_bug` occurs here:
```rust
impl<'tcx> MirPass<'tcx> for ElaborateDrops {
    fn run_pass(&self, tcx: TyCtxt<'tcx>, src: MirSource<'tcx>, body: &mut BodyAndCache<'tcx>) {
        debug!("elaborate_drops({:?} @ {:?})", src, body.span);

        let def_id = src.def_id();
        let param_env = tcx.param_env(src.def_id()).with_reveal_all();
        let move_data = match MoveData::gather_moves(body, tcx, param_env) {
            Ok(move_data) => move_data,
            Err((move_data, _)) => {
                tcx.sess.delay_span_bug(
                    body.span,
                    "No `move_errors` should be allowed in MIR borrowck",
                );
                move_data
            }
        };
```

cc @oli-obk @matthewjasper

I guess we need to track down where `error[E0508]: cannot move out of type `[S; 1]`, a non-copy array` is emitted and make that code also look at the MIR of promoteds.

We can't avoid that code from getting promoted without getting into an unfun game of whack-a-mole in the promotion code.

--- a/src/librustc_mir/borrow_check/mod.rs
+++ b/src/librustc_mir/borrow_check/mod.rs
@@ -180,11 +180,14 @@ fn do_mir_borrowck<'a, 'tcx>(
     let location_table = &LocationTable::new(&body);
 
     let mut errors_buffer = Vec::new();
-    let (move_data, move_errors): (MoveData<'tcx>, Option<Vec<(Place<'tcx>, MoveError<'tcx>)>>) =
+    let (move_data, move_errors): (MoveData<'tcx>, Vec<(Place<'tcx>, MoveError<'tcx>)>) =
         match MoveData::gather_moves(&body, tcx, param_env) {
-            Ok(move_data) => (move_data, None),
-            Err((move_data, move_errors)) => (move_data, Some(move_errors)),
+            Ok(move_data) => (move_data, Vec::new()),
+            Err((move_data, move_errors)) => (move_data, move_errors),
         };
+    let promoted_errors = promoted
+        .iter_enumerated()
+        .map(|(idx, body)| (idx, MoveData::gather_moves(&body, tcx, param_env)));
 
     let mdpe = MoveDataParamEnv { move_data, param_env };
 
@@ -264,6 +267,41 @@ fn do_mir_borrowck<'a, 'tcx>(
         _ => true,
     };
 
+    for (idx, move_data_results) in promoted_errors {
+        let promoted_body = &promoted[idx];
+        let dominators = promoted_body.dominators();
+
+        if let Err((move_data, move_errors)) = move_data_results {
+            let mut promoted_mbcx = MirBorrowckCtxt {
+                infcx,
+                body: promoted_body,
+                mir_def_id: def_id.to_def_id(),
+                move_data: &move_data,
+                location_table: &LocationTable::new(promoted_body),
+                movable_generator,
+                locals_are_invalidated_at_exit,
+                access_place_error_reported: Default::default(),
+                reservation_error_reported: Default::default(),
+                reservation_warnings: Default::default(),
+                move_error_reported: BTreeMap::new(),
+                uninitialized_error_reported: Default::default(),
+                errors_buffer,
+                regioncx: regioncx.clone(),
+                used_mut: Default::default(),
+                used_mut_upvars: SmallVec::new(),
+                borrow_set: borrow_set.clone(),
+                dominators,
+                upvars: Vec::new(),
+                local_names: IndexVec::from_elem(None, &promoted_body.local_decls),
+                region_names: RefCell::default(),
+                next_region_name: RefCell::new(1),
+                polonius_output: None,
+            };
+            promoted_mbcx.report_move_errors(move_errors);
+            errors_buffer = promoted_mbcx.errors_buffer;
+        };
+    }
+
     let dominators = body.dominators();
 
     let mut mbcx = MirBorrowckCtxt {
@@ -301,9 +339,7 @@ fn do_mir_borrowck<'a, 'tcx>(
         borrows: flow_borrows,
     };
 
-    if let Some(errors) = move_errors {
-        mbcx.report_move_errors(errors);
-    }
+    mbcx.report_move_errors(move_errors);
 
     dataflow::visit_results(
         &body,

--- a/src/test/ui/borrowck/move-error-in-promoted-2.rs
+++ b/src/test/ui/borrowck/move-error-in-promoted-2.rs
@@ -1,0 +1,10 @@
+// Regression test for #70934
+
+struct S;
+
+fn foo() {
+    &([S][0],);
+    //~^ ERROR cannot move out of type `[S; 1]`
+}
+
+fn main() {}

--- a/src/test/ui/borrowck/move-error-in-promoted-2.stderr
+++ b/src/test/ui/borrowck/move-error-in-promoted-2.stderr
@@ -1,0 +1,12 @@
+error[E0508]: cannot move out of type `[S; 1]`, a non-copy array
+  --> $DIR/move-error-in-promoted-2.rs:6:7
+   |
+LL |     &([S][0],);
+   |       ^^^^^^
+   |       |
+   |       cannot move out of here
+   |       move occurs because value has type `S`, which does not implement the `Copy` trait
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0508`.

--- a/src/test/ui/borrowck/move-error-in-promoted.rs
+++ b/src/test/ui/borrowck/move-error-in-promoted.rs
@@ -1,0 +1,17 @@
+// Regression test for #70934
+
+fn f() {
+    const C: [S2; 1] = [S2];
+    let _ = S1(C[0]).clone();
+    //~^ ERROR cannot move out of type `[S2; 1]`
+}
+
+#[derive(Clone)]
+struct S1(S2);
+
+#[derive(Clone)]
+struct S2;
+
+fn main() {
+    f();
+}

--- a/src/test/ui/borrowck/move-error-in-promoted.stderr
+++ b/src/test/ui/borrowck/move-error-in-promoted.stderr
@@ -1,0 +1,12 @@
+error[E0508]: cannot move out of type `[S2; 1]`, a non-copy array
+  --> $DIR/move-error-in-promoted.rs:5:16
+   |
+LL |     let _ = S1(C[0]).clone();
+   |                ^^^^
+   |                |
+   |                cannot move out of here
+   |                move occurs because value has type `S2`, which does not implement the `Copy` trait
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0508`.

